src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectLockFileWatcher.cs

```diff
@@ -0,0 +1,165 @@
+        {
+            this.EnsureInitialized();
+        }
+        
+        /// <summary>
+        /// Initialize the watcher.
+        /// </summary>
+        protected override void Initialize()
+        {
+            _treeWatcher = _fileSystemTreeProvider.Tree.LinkTo(new ActionBlock<IProjectVersionedValue<IProjectTreeSnapshot>>(new Action<IProjectVersionedValue<IProjectTreeSnapshot>>(this.ProjectTree_ChangedAsync)));
+        }
+
+        /// <summary>
+        /// Called on changes to the project tree.
+        /// </summary>
+        internal void ProjectTree_ChangedAsync(IProjectVersionedValue<IProjectTreeSnapshot> treeSnapshot)
+        {
+            var newTree = treeSnapshot.Value.Tree;
+            if (newTree == null)
+            {
+                return;
+            }
+
+            // If there' no project.json in the project, there's nothing to watch.
+            IProjectTree projectJsonNode = FindProjectJsonNode(newTree);
+            if (projectJsonNode == null)
+            {
+                // project.json may have been deleted.
+                UnregisterFileWatcherIfAny();
+                return;
+            }
+
+            var projectDirectory = Path.GetDirectoryName(_projectServices.Project.FullPath);
+            var projectLockJsonFilePath = Path.ChangeExtension(PathHelper.Combine(projectDirectory, projectJsonNode.Caption), ".lock.json");
+
+            // project.json may have been renamed to {projectName}.project.json. In that case change the file watcher.
+            if (!PathHelper.IsSamePath(projectLockJsonFilePath, _fileBeingWatched))
+            {
+                UnregisterFileWatcherIfAny();
+                RegisterFileWatcher(projectLockJsonFilePath);
+                _fileBeingWatched = projectLockJsonFilePath;
+            }
+        }
+
+        private IProjectTree FindProjectJsonNode(IProjectTree newTree)
+        {
+            IProjectTree projectJsonNode;
+            if (newTree.TryFindImmediateChild("project.json", out projectJsonNode))
+            {
+                return projectJsonNode;
+            }
+
+            var projectName = Path.GetFileNameWithoutExtension(_projectServices.Project.FullPath);
+            if (newTree.TryFindImmediateChild($"{projectName}.project.json", out projectJsonNode))
+            {
+                return projectJsonNode;
+            }
+
+            return null;
+        }
+
+        private void RegisterFileWatcher(string projectLockJsonFilePath)
+        {
+            IVsFileChangeEx fileChangeService = _serviceProvider.GetService<IVsFileChangeEx, SVsFileChangeEx>();
+            if (fileChangeService != null)
+            {
+                int hr = fileChangeService.AdviseFileChange(projectLockJsonFilePath, (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Time | _VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Add | _VSFILECHANGEFLAGS.VSFILECHG_Del), this, out _filechangeCookie);
+                ErrorHandler.ThrowOnFailure(hr);
```

Comment:
Yes that was unadvise. I specifically tested the case of missing file and then restore adding the project.lock.json